### PR TITLE
Add circular padding for artifact-free StableDiffusionPanoramaPipeline

### PR DESCRIPTION
**Pull request by EvgenyKashin**
## Artifact-Free Panorama Generation with circular padding
This pull request addresses the issue mentioned in https://github.com/huggingface/diffusers/issues/4024. The proposed solution aims to achieve artifact-free panorama generation by seamlessly **aligning the content on the left and right sides of the generated image**, creating a proper “panorama”, viewable in 360 panoramas viewers.

In the existing MultiDiffusion approach, the image is cropped from left to right with a specified stride. To enable circular padding and support the desired 360-degree scenario, this PR introduces additional crops after the rightmost point of the image. **Circular padding allows the model to “see” the transition from the rightmost part of the image to the leftmost part, maintaining consistency in 360-sense**.

Furthermore, modifications are made to the decoding of latents in the case of StableDiffusion. Without these modifications, the decoded latents would not match in the RGB space, even if they align well in the latent space. Thus, **the decoding of latents also incorporates circular padding to ensure consistent results**.

A panorama image generated **without circular padding**: image

A panorama image generated **with circular padding**: image

To maintain backward compatibility and preserve the previous behaviour, the proposed circular padding functionality is implemented as an optional argument “circular_padding”, which is set to False by default.

Usage example:
```python
import torch
from diffusers import StableDiffusionPanoramaPipeline, DDIMScheduler

model_ckpt = "stabilityai/stable-diffusion-2-base"
scheduler = DDIMScheduler.from_pretrained(model_ckpt, subfolder="scheduler")
pipe = StableDiffusionPanoramaPipeline.from_pretrained(model_ckpt, scheduler=scheduler, torch_dtype=torch.float16)
pipe = pipe.to("cuda")

prompt = "an indoor equirectangular panorama of a living room"

image = pipe(prompt, width=1024, height=512,
    circular_padding=True).images[0]
```

### CODE REVIEW

**Comment by EvgenyKashin** [comment]
Visualization of a concatenation of the most right and the most left parts of a generated panorama images: No circular padding (default):
 image

Circular padding, but without decoding padding in the VAE (`decode_latents_with_padding`):
image

Circular padding full (`circular_padding=True`):
image

**Comment by EvgenyKashin** [comment]
In general, it's an easy fix for the stitching artifact problem, but maybe I've missed an easier way to do it with the current pipeline. Perhaps this is too narrow case to add this to the main. 

I've also noticed [tests](https://github.com/huggingface/diffusers/blob/main/tests/pipelines/stable_diffusion/test_stable_diffusion_panorama.py) for this pipeline that I break with my changes (even with default `circular_padding=False` option, that is just the same as current main in my local tests):
`18 failed, 5 passed, 5 skipped, 2 warnings`

I've tried "cuda" and "cpu", different `batch_size` and `view_batch_size`, it works fine. I couldn't understand why it's failing tests yet.

**Comment by EvgenyKashin** [comment]
@omerbt @sayakpaul @Isotr0py guys, what do you think?

**Comment by sayakpaul** [comment]
I will defer to @omerbt for this one. But I like what I see :)

**Comment by omerbt** [comment]
Looks great!

**Comment by sayakpaul** [comment]
Thanks @omerbt!

@EvgenyKashin, let's make sure the tests on the CI pass. Then I think we can approach merging.

**Comment by EvgenyKashin** [comment]
Thanks, guys. @omerbt I've fixed one place in the code. Basically, I decided to save one line originally, but one more check was required for small images case (like in test cases). I've also added a few simple test cases for circular padding.

**Comment by EvgenyKashin** [comment]
Sorry, originally tagged the wrong person, @sayakpaul

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_panorama.py
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_panorama.py:581]: Maybe explain a bit what circular padding is and how it helps?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_panorama.py
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_panorama.py:764]: Works for me for now. Ideally, I would have loved to move out the VAE decode part from `decode_latents_with_padding()` to unify this part of the workflow a bit. But okay for now.

File: tests/pipelines/stable_diffusion/test_stable_diffusion_panorama.py
Comment (sayakpaul) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_panorama.py:297]: I think it's okay to remove this slow test.

**Review by sayakpaul — APPROVED** [review]
Looks great to me! Thank you. 

Left some minor nits. One thing we could do is add a note about `circular_padding` to the panorama docs: https://github.com/huggingface/diffusers/blob/main/docs/source/en/api/pipelines/panorama.mdx.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_panorama.py
Comment (EvgenyKashin) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_panorama.py]: Done

File: tests/pipelines/stable_diffusion/test_stable_diffusion_panorama.py
Comment (EvgenyKashin) [reply tests/pipelines/stable_diffusion/test_stable_diffusion_panorama.py]: Removed

**Comment by EvgenyKashin** [comment]
Could we merge it?

File: docs/source/en/api/pipelines/panorama.mdx
Comment (sayakpaul) [inline docs/source/en/api/pipelines/panorama.mdx:72]: If possible would be also nice to link the comparison with and without circular padding. This way, the readers can spot-check the effectivity.

Comment (EvgenyKashin) [reply docs/source/en/api/pipelines/panorama.mdx:72]: Do you mean adding a text description of “normal” multi-diffusion, or adding a link to an image example of the results?

Comment (sayakpaul) [reply docs/source/en/api/pipelines/panorama.mdx:72]: > adding a link to an image example of the results?

Yeah just noting the difference with a comparative example like the one you showed in the PR description.

Comment (EvgenyKashin) [reply docs/source/en/api/pipelines/panorama.mdx:72]: That was a good idea indeed

**Comment by sayakpaul** [comment]
Will merge after the CI passes.

**Comment by sayakpaul** [comment]
Thanks very much for your valuable contributions!

**Pull request merged**
